internal/dag/updatable_dag.go

```diff
@@ -0,0 +1,249 @@
+	}
+	d.nodes[node.Identifier()] = node
+	return nil
+}
+
+// AddOrUpdateNodes adds new nodes or updates the existing ones with the same
+// identifier.
+func (d *MapUpdatableDag) AddOrUpdateNodes(nodes ...Node) {
+	for _, node := range nodes {
+		node.AddParentConstraints(d.nodes[node.Identifier()].GetParentConstraints())
+		d.nodes[node.Identifier()] = node
+	}
+}
+
+// NodeExists checks whether a node exists.
+func (d *MapUpdatableDag) NodeExists(identifier string) bool {
+	_, exists := d.nodes[identifier]
+	return exists
+}
+
+// NodeNeighbors returns a node's neighbors.
+func (d *MapUpdatableDag) NodeNeighbors(identifier string) ([]Node, error) {
+	if _, ok := d.nodes[identifier]; !ok {
+		return nil, errors.Errorf("node %s does not exist", identifier)
+	}
+	return d.nodes[identifier].Neighbors(), nil
+}
+
+// TraceNode returns a node's neighbors and all transitive neighbors using depth
+// first search.
+func (d *MapUpdatableDag) TraceNode(identifier string) (map[string]Node, error) {
+	tree := map[string]Node{}
+	if err := d.traceNode(identifier, tree); err != nil {
+		return nil, err
+	}
+	return tree, nil
+}
+
+func (d *MapUpdatableDag) traceNode(identifier string, tree map[string]Node) error {
+	if d.nodes[identifier] == nil {
+		return errors.New("missing node in tree")
+	}
+	for _, n := range d.nodes[identifier].Neighbors() {
+		// if we have already visited this neighbor, then we have already
+		// visited its neighbors, so we can skip.
+		if _, ok := tree[n.Identifier()]; ok {
+			continue
+		}
+		tree[n.Identifier()] = n
+		if err := d.traceNode(n.Identifier(), tree); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetNode returns a node in the dag.
+func (d *MapUpdatableDag) GetNode(identifier string) (Node, error) {
+	if _, ok := d.nodes[identifier]; !ok {
+		return nil, errors.Errorf("node %s does not exist", identifier)
+	}
+	return d.nodes[identifier], nil
+}
+
+// AddEdges adds edges to the graph.
+func (d *MapUpdatableDag) AddEdges(edges map[string][]Node) ([]Node, error) {
+	var missing []Node
+	for f, ne := range edges {
+		for _, e := range ne {
+			implied, err := d.AddEdge(f, e)
+			if implied {
+				missing = append(missing, e)
+			}
+
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return missing, nil
+}
+
+// AddEdge adds an edge to the graph and returns if we need to check for updates.
+func (d *MapUpdatableDag) AddEdge(from string, to Node) (bool, error) {
+	if _, ok := d.nodes[from]; !ok {
+		return false, errors.Errorf("node %s does not exist", to)
+	}
+
+	implied := false
+	orgTo, ok := d.nodes[to.Identifier()]
+	if !ok {
+		implied = true
+		if err := d.AddNode(to); err != nil {
+			return implied, err
+		}
+	} else if !isValidConstraints(orgTo, to) { // check if upgrade is needed
+		err := d.nodes[from].AddNeighbors(to)
+		n := d.nodes[to.Identifier()]
+		n.AddParentConstraints(to.GetParentConstraints())
+
+		return true, err
```

Comment:
Am I following correctly that the only difference between this block and the following block is that it always returns `implied == true`?

Reply:
if the new node we want to add already exists in the DAG, and asks for a different constraint than what we currently satisfy, we return `implied == true` so that we can check whether there is a valid version which satisfies both existing packages' constraints and the new constraint